### PR TITLE
Feature: Get Folder Content

### PR DESCRIPTION
**Pull request by YehorPytomets**
### Changes introduced by this PR:
* Implement navigation between folders of `ListExplorer` page.
* Open folder by double-clicking it.
* Refactor `Router` implementation.

### CODE REVIEW

File: webclient/app/actions/get-root-folder-action.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+export default class GetRootFolderAction extends Action {
+
+  async apply(stateManager) {
+    const mutator = new FolderLoadingMutator(true);
+    stateManager.mutate(mutator);
+
+    return stateManager.apiService.getRootFolder()
+      .then((folderResponseJson) => {
+        const mutator = new GetRootFolderMutator(folderResponseJson);
```
Comment (dkalinin) [inline webclient/app/actions/get-root-folder-action.js:14]: - [ ] Please rename GetRootFolderMutator.

File: webclient/app/actions/get-root-folder-action.js [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+        stateManager.mutate(mutator);
+      })
+      .catch((error) => {
+        const mutator = new GetRootFolderLoadingErrorMutator(error);
```
Comment (dkalinin) [inline webclient/app/actions/get-root-folder-action.js:18]: - [ ] Please rename.

File: webclient/app/actions/get-folder-action.js [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+   * @inheritdoc
+   * @param {StateManager} stateManager - state manager with target
+   */
+  async apply(stateManager) {
+    const mutator = new FolderLoadingMutator(true);
+    stateManager.mutate(mutator);
+
+    return stateManager.apiService.getFolder(this._folderId)
+      .then((folderResponseJson) => {
+        const mutator = new GetFolderMutator(folderResponseJson);
```
Comment (dkalinin) [inline webclient/app/actions/get-folder-action.js:38]: - [ ] Please rename mutator.

File: webclient/app/app.js [hunk 4]
```diff
@@ -68,37 +75,74 @@ export default class Application extends Component {
...
+        return new NotFoundPage(rootElement, {
+          titleService,
+        });
+      })
+      .withHashChangeHandler((staticPart, dynamicPart) => {
+        const routeChangedAction = new RouteChangedAction(staticPart, dynamicPart);
+        stateManager.dispatchAction(routeChangedAction);
+      })
+      .withInitRouterHandler((router) => {
+        stateManager.onStateChanged('location', (event) => {
```
Comment (dkalinin) [inline webclient/app/app.js:135]: - [ ] Router should not listen to state.

File: webclient/app/components/file-list/file-list.js [hunk 5]
```diff
@@ -49,7 +59,7 @@ export default class FileList extends Component {
    * @return {null|FolderComponent|FileComponent} - instantiated component.
    */
   _createListItemComponent(container, listItem) {
-    const {FILE, FOLDER} = ListItemTypes;
+    const {FILE, FOLDER} = ListItemType;
     const listItemTypeCreatorMap = {
       [FILE]: () => new FileComponent(container, {listItem}),
       [FOLDER]: () => new FolderComponent(container, {listItem}),
```
Comment (dkalinin) [inline webclient/app/components/file-list/file-list.js:65]: - [ ] Please don't return null.

File: webclient/app/components/file-list/list-item-component.js [hunk 6]
```diff
@@ -14,6 +15,27 @@ export default class ListItemComponent extends Component {
...
+    this._onDoubleClickHandlers.executeHandlers(event);
+  });
+  }
+
+  /**
+   * Adds on-double-click-event handler.
+   *
+   * @param {function(Event): void} handler - handler for double click event.
+   */
+  onDoubleClick(handler) {
```
Comment (dkalinin) [inline webclient/app/components/file-list/list-item-component.js:37]: - [ ] This works for folder only.

File: webclient/app/models/file-model.js [hunk 7]
```diff
@@ -1,4 +1,7 @@
 import ListItemModel from './list-item-model.js';
 
+/**
+ * Model to store folder data.
```
Comment (dkalinin) [inline webclient/app/models/file-model.js:4]: file

File: webclient/app/mutators/change-location-mutator.js [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+   * @inheritdoc
+   * @param {string} location - location of the app.
+   */
+  constructor(location) {
+    super();
+    this._location = location;
+  }
+
+  /**
+   * Helps router to prevent rendering page if location didn't change.
```
Comment (dkalinin) [inline webclient/app/mutators/change-location-mutator.js:24]: - [ ] It just changes state.

File: webclient/app/pages/file-list-explorer.js [hunk 9]
```diff
@@ -121,15 +121,16 @@ export default class FileListExplorer extends StateAwareComponent {
...
 
     this.onStateChanged('listItems', (event) => {
       listComponent.renderListItems(event.detail.state.listItems);
     });
 
     this.onStateChanged('isListItemsLoading', (event) => {
-      const classList = this.rootElement.querySelector('[data-type="loader"]').classList;
-      const loaderClass = 'loader';
+      const classList = this.rootElement.querySelector('[data-type="folder-content-loader"]').classList;
+      const loaderClass = 'folder-content-loader';
```
Comment (dkalinin) [inline webclient/app/pages/file-list-explorer.js:134]: - [ ] This should be moved to setter of isLoading field.

File: webclient/app/pages/file-list-explorer.js [hunk 10]
```diff
@@ -160,11 +203,56 @@ export default class FileListExplorer extends StateAwareComponent {
...
+  _handleRootFolderReceived(rootFolderId) {
+    this._onRootFolderReceived.executeHandlers(rootFolderId);
+  }
+
+  /**
+   * Registers handler for event when root folder meta data is received.
+   *
+   * @param {function(rootFolderId: string)} handler - handler to register.
+   */
+  onRootFolderReceived(handler) {
```
Comment (dkalinin) [inline webclient/app/pages/file-list-explorer.js:246]: - [ ] Please rename.

File: webclient/app/pages/file-list-explorer.js [hunk 11]
```diff
@@ -160,11 +203,56 @@ export default class FileListExplorer extends StateAwareComponent {
...
+   * Registers resource-not-found-error handler.
+   *
+   * @param {function()} handler - handler to register.
+   */
+  onResourceNotFoundError(handler) {
+    this._onResourceNotFoundErrorHandlers.addEventHandler(handler);
+  }
+
+  /**
+   * Handles resource-not-found error.
```
Comment (dkalinin) [inline webclient/app/pages/file-list-explorer.js:232]: - [ ] root folder received.

File: webclient/app/pages/file-list-explorer.js [hunk 12]
```diff
@@ -160,11 +203,56 @@ export default class FileListExplorer extends StateAwareComponent {
...
+  onRootFolderReceived(handler) {
+    this._onRootFolderReceived.addEventHandler(handler);
+  }
+
+  /**
+   * Registers handler, that executes when double click was produced on item.
+   *
+   * @param {function(itemId: string): void} handler - handler to register.
+   */
+  addDoubleClickListItemHandler(handler) {
```
Comment (dkalinin) [inline webclient/app/pages/file-list-explorer.js:255]: - [ ] Please rename.

**Comment by YehorPytomets** [comment]
@dkalinin , PR is ready to review.

### What was done:
* Resolve tests conflicts after merge of `feature/state` PR.
* Change router implementation.
* Add GetRootFolder request.
* Handle statechanged.* event in FileListExplorer.
* Add new tests for actions, mutators, breadcrumbs, apiService, router.
* Fix comments of previous review and do minor fixes.

File: webclient/app/components/breadcrumbs.js [hunk 13]
```diff
@@ -54,24 +54,69 @@ export default class Breadcrumbs extends Component {
...
+  _renderTrailText(text = '', isErrorText = false) {
+    const textContainer = this._getTextContainer();
+
+    if (isErrorText) {
+      textContainer.classList.add('error-text');
+    } else {
+      textContainer.classList.remove('error-text');
+    }
+
+    textContainer.textContent = text ? `/ ${text}` : '';
```
[the remark this answers is not in the document]
Comment (YehorPytomets) [reply webclient/app/components/breadcrumbs.js:70]: This will be done in separate PR: `Refactor: Breadcrumbs`

**Comment by YehorPytomets** [comment]
@dkalinin , PR is ready to review.

File: webclient/tests/app/actions/change-location-parameters-action-tests.js [hunk 14]
```diff
@@ -14,7 +14,12 @@ export default (() => {
       parameter: '0000',
     };
     const mockState = new State();
-    const stateManager = new MockStateManger(mockState, assert);
+    const stateManager = {
+      mutate(mutator) {
+        assert.step(`Called mutator: ${mutator.constructor.name}.`);
+        mockState.locationParams = locationParameters;
```
Comment (dkalinin) [inline webclient/tests/app/actions/change-location-parameters-action-tests.js:20]: - [ ] Please replace this with assert. And remove mockState.

File: webclient/tests/app/actions/get-root-folder-action-tests.js [hunk 15]
```diff
@@ -12,106 +9,76 @@ export default (() => {
...
-      id: 'mock-id',
+    assert.expect(8);
+
+    const stateManager = {
+      apiService: {
+        getRootFolder: async () => {
+          assert.step('Should call ApiService.getRootFolder().');
+
+          return {
+            id: 'mock-id',
```
Comment (dkalinin) [inline webclient/tests/app/actions/get-root-folder-action-tests.js:20]: - [ ] Please move to variable.

File: webclient/app/actions/route-changed-action.js [hunk 16]
```diff
@@ -26,13 +26,14 @@ export default class RouteChangedAction extends Action {
...
-        }
-      });
+    if (stateManager.state.location !== this.staticPart) {
+      const locationAction = new ChangeLocationAction(staticPart);
+      await stateManager.dispatchAction(locationAction)
+    }
+
+    if (Object.keys(dynamicPart).length) {
+      const locationParametersAction = new ChangeLocationParametersAction(dynamicPart);
+      await stateManager.dispatchAction(locationParametersAction);
```
Comment (dkalinin) [inline webclient/app/actions/route-changed-action.js:36]: - [ ] Should be called in any case if static part is changed.

File: webclient/app/actions/route-changed-action.js [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+  async apply(stateManager) {
+    const {staticPart, dynamicPart} = this;
+
+    if (stateManager.state.location !== this.staticPart) {
+      const locationAction = new ChangeLocationAction(staticPart);
+      await stateManager.dispatchAction(locationAction)
+    }
+
+    if (Object.keys(dynamicPart).length) {
+      const locationParametersAction = new ChangeLocationParametersAction(dynamicPart);
```
Comment (dkalinin) [inline webclient/app/actions/route-changed-action.js:35]: - [ ] Let's remove ChangeLocationParametersAction and ChangeLocationAction and call mutators instead.

**Comment by YehorPytomets** [comment]
@dkalinin , PR is ready to review.

File: webclient/app/actions/route-changed-action.js [hunk 18]
```diff
@@ -16,24 +28,26 @@ export default class RouteChangedAction extends Action {
...
+    if (stateManager.state.location !== _staticPart) {
+      const locationMutator = new LocationMutator(_staticPart);
+      stateManager.mutate(locationMutator);
+      stateManager.mutate(locationParametersMutator);
+    } else {
 
-    if (Object.keys(dynamicPart).length) {
-      const locationParametersAction = new ChangeLocationParametersAction(dynamicPart);
-      await stateManager.dispatchAction(locationParametersAction);
+      if (Object.keys(_dynamicPart).length) {
```
Comment (dkalinin) [inline webclient/app/actions/route-changed-action.js]: Let's avoid else-if conditions. Just return in previous if block.
https://softwareengineering.stackexchange.com/a/206867

**Pull request merged**